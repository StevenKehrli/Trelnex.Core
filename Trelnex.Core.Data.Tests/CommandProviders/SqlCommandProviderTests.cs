using Azure.Core;
using Azure.Identity;
using Microsoft.Data.SqlClient;
using Microsoft.Extensions.Configuration;

namespace Trelnex.Core.Data.Tests.CommandProviders;

[Ignore("Requires a SQL server.")]
public class SqlCommandProviderTests : CommandProviderTests
{
    private readonly string _scope = "https://database.windows.net/.default";

    private TokenCredential _tokenCredential = null!;
    private string _connectionString = null!;
    private string _tableName = null!;

    [OneTimeSetUp]
    public async Task TestFixtureSetup()
    {
        // This method is called once prior to executing any of the tests in the fixture.

        // create the test configuration
        var configuration = new ConfigurationBuilder()
            .AddJsonFile("appsettings.json", optional: true, reloadOnChange: true)
            .AddJsonFile("appsettings.User.json", optional: true, reloadOnChange: true)
            .Build();

        var providerConfiguration = configuration.GetSection("SqlCommandProviders").Get<SqlCommandProviderConfiguration>()!;

        var scsBuilder = new SqlConnectionStringBuilder()
        {
            DataSource = providerConfiguration.DataSource,
            InitialCatalog = providerConfiguration.InitialCatalog,
            Encrypt = true,
        };

        _connectionString = scsBuilder.ConnectionString;
        _tableName = providerConfiguration.TableName;

        // create the command provider
        _tokenCredential = new DefaultAzureCredential();

        var sqlClientOptions = new SqlClientOptions(
            TokenCredential: _tokenCredential,
            Scope: _scope,
            DataSource: providerConfiguration.DataSource,
            InitialCatalog: providerConfiguration.InitialCatalog,
            TableNames: [ providerConfiguration.TableName ]
        );

        var factory = await SqlCommandProviderFactory.Create(
            sqlClientOptions);

        _commandProvider = factory.Create<ITestItem, TestItem>(
            providerConfiguration.TableName,
            "test-item",
            TestItem.Validator,
            CommandOperations.All);
    }

    [TearDown]
    public void TestCleanup()
    {
        // This method is called after each test has run.
        using var sqlConnection = new SqlConnection(_connectionString);

        var tokenRequestContext = new TokenRequestContext([ _scope ]);
        sqlConnection.AccessToken = _tokenCredential.GetToken(tokenRequestContext, default).Token;

        sqlConnection.Open();

        var cmdText = $"DELETE FROM [{_tableName}-events]; DELETE FROM [{_tableName}];";
        var sqlCommand = new SqlCommand(cmdText, sqlConnection);

        sqlCommand.ExecuteNonQuery();
    }

    /// <summary>
    /// Represents the configuration properties for SQL command providers.
    /// </summary>
    private record SqlCommandProviderConfiguration(
        string DataSource,
        string InitialCatalog,
        string TableName);
}
