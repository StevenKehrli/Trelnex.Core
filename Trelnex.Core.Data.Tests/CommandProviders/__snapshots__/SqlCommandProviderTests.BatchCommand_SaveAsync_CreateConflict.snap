{
  "saved": [
    {
      "HttpStatusCode": "FailedDependency",
      "ReadResult": null
    },
    {
      "HttpStatusCode": "InternalServerError",
      "ReadResult": null
    }
  ],
  "read": [
    {
      "Item": {
        "Id": "a0ac8a58-2d76-40d2-9f14-1dd5543e8429",
        "PartitionKey": "ae9d5360-1e01-480a-9465-4da82ee882b7",
        "TypeName": "test-item",
        "CreatedDate": "2025-01-10T17:09:41.346422Z",
        "UpdatedDate": "2025-01-10T17:09:41.346422Z",
        "DeletedDate": null,
        "IsDeleted": null,
        "ETag": "B49E60F7-724D-4E06-8528-5382A0D4825C",
        "Message": "Message #0"
      }
    }
  ]
}
